sdks/go/pkg/beam/io/datastoreio/datastore.go

```diff
@@ -0,0 +1,195 @@
+package datastoreio
+
+import (
+	"reflect"
+	"github.com/apache/beam/sdks/go/pkg/beam"
+	"sort"
+	"google.golang.org/api/iterator"
+	"encoding/json"
+	"fmt"
+	"cloud.google.com/go/datastore"
+	"context"
+	"math"
+	"github.com/apache/beam/sdks/go/pkg/beam/core/runtime"
+	"strconv"
+)
+
+const (
+	scatterPropertyName = "__scatter__"
+)
+
+// Read reads all rows from the given kind. The kind must have a schema  compatible with the given type, t, and Read
+// returns a PCollection<t>. You must also register your type with runtime.RegisterType which allows you to implement
+// datastore.PropertyLoadSaver
+//
+//
+// Example:
+// type Item struct {}
+// itemKey = runtime.RegisterType(reflect.TypeOf((*Item)(nil)).Elem())
+//
+// datastoreio.Read(s, "project", "Item", 256, reflect.TypeOf(Item{}), itemKey)
+func Read(s beam.Scope, project, kind string, shards int, t reflect.Type, typeKey string) beam.PCollection {
+	s = s.Scope("datastore.Read")
+	return query(s, project, kind, shards, t, typeKey)
+}
+
+func query(s beam.Scope, project, kind string, shards int, t reflect.Type, typeKey string) beam.PCollection {
+	imp := beam.Impulse(s)
+	ex := beam.ParDo(s, &splitQueryFn{Project: project, Kind: kind, Shards: shards}, imp)
+	g := beam.GroupByKey(s, ex)
+	return beam.ParDo(s, &queryFn{Project: project, Kind: kind, Type: typeKey}, g, beam.TypeDefinition{Var: beam.XType, T: t})
+}
+
+type splitQueryFn struct {
+	Project string `json:"project"`
+	Kind    string `json:"kind"`
+	Shards  int    `json:"shards"`
+}
+
+// BoundedQuery represents a datastore Query with a bounded key range between [Start, End)
+type BoundedQuery struct {
+	Start *datastore.Key `json:"start"`
+	End   *datastore.Key `json:"end"`
+}
+
+func (s *splitQueryFn) ProcessElement(ctx context.Context, _ []byte, emit func(k string, val string)) error {
+	client, err := datastore.NewClient(ctx, s.Project)
+	if err != nil {
+		return err
+	}
+	defer client.Close()
+
+	splits := []*datastore.Key{}
+	iter := client.Run(ctx, datastore.NewQuery(s.Kind).Order(scatterPropertyName).Limit((s.Shards - 1) * 32).KeysOnly())
+	for {
+		k, err := iter.Next(nil)
+		if err != nil {
+			if err == iterator.Done {
+				break
+			}
+			return err
+		}
+		splits = append(splits, k)
+	}
+	sort.Slice(splits, func(i, j int) bool {
+		return keyLessThan(splits[i], splits[j])
+	})
+
+	splitKeys := getSplits(splits, s.Shards)
+
+	queries := []*BoundedQuery{}
+	var lastKey *datastore.Key
+	for _, k := range splitKeys {
+		q := BoundedQuery{End: k}
+		if lastKey != nil {
+			q.Start = lastKey
+		}
+		queries = append(queries, &q)
+		lastKey = k
+	}
+	queries = append(queries, &BoundedQuery{End: lastKey})
```

Comment:
Shouldn't the last query be {Start: lastKey}? Won't we otherwise have an enormous query here?